src/ja/server/dispatcher/dispatcher.py

```diff
@@ -22,38 +25,79 @@ def __init__(self, proxy_factory: WorkerProxyFactoryBase):
         self._proxy_factory = proxy_factory
         self._previous_statuses: Dict[str, JobStatus] = dict()
 
-    def set_distribution(self, job_distribution: JobDistribution) -> None:
+    def set_distribution(self, job_distribution: JobDistribution) -> List[WorkMachine]:
         """!
         Distributes the newly added jobs to the work machines.
         and potentially pauses jobs that need to be paused.
 
         @param job_distribution the new job distribution.
+        @return a list of all work machines that the server has lost connection to.
         """
         logger.debug("Dispatching jobs")
         new_statuses: Dict[str, JobStatus] = dict()
+        self._lost_work_machines: List[WorkMachine] = []
         for job_entry in job_distribution:
             job = job_entry.job
 
             if job.status == JobStatus.QUEUED:
                 logger.debug("Job %s queued." % job.uid)
                 continue
+            if job.status == JobStatus.RUNNING:
+                continue
 
             work_machine = job_entry.assigned_machine
             logger.debug("Job %s with status %s on %s." % (job.uid, job.status.name, work_machine.uid))
+            proxy = self._proxy_factory.get_proxy(work_machine)
+            try:
+                proxy.check_connection()
+            except SSHException:
+                self._timeout(job_entry)
```

Comment:
Why are you checking the connection to the work machine separately for each job entry?
In the code below there will be a timeout check anyways, right?

Reply:
If the job status didn't change, there won't be a check. So if the server loses connection to a work machine that's running a job, the server keeps storing it as a running job indefinitely.